### PR TITLE
Support SHT_NOTE sections

### PR DESCRIPTION
**Pull request by gbenson**
This commit adds a new NoteSection class with an iter_notes iterator which operates in much the same way as NoteSegment's iter_notes.

### CODE REVIEW

File: elftools/elf/sections.py [hunk 1]
```diff
@@ -174,3 +174,15 @@ def iter_symbols(self):
...
+    """ ELF NOTE section. Knows how to parse notes.
+    """
+    def __init__(self, header, name, stream, elffile):
+        super(NoteSection, self).__init__(header, name, stream)
+        self.elffile = elffile
+
+    def iter_notes(self):
+        """ Iterates the list of notes in the segment.
+        """
+        return self.elffile._iter_notes(self['sh_offset'], self['sh_size'])
```
Comment (eliben) [inline elftools/elf/sections.py]: Why not do the parsing here, in NoteSection?

File: elftools/elf/segments.py [hunk 2]
```diff
@@ -102,36 +101,9 @@ class NoteSegment(Segment):
...
-                                              self.stream,
-                                              offset)
-            elif note['n_type'] == 'NT_GNU_BUILD_ID':
-                note['n_desc'] = ''.join('%.2x' % ord(b) for b in desc_data)
-            else:
-                note['n_desc'] = desc_data
-            offset += roundup(note['n_descsz'], 2)
-            note['n_size'] = offset - note['n_offset']
-            yield note
+        return self.elffile._iter_notes(self['p_offset'], self['p_filesz'])
```
Comment (eliben) [inline elftools/elf/segments.py]: Ah I see... I'd prefer to have it factored into a separate module, like elf/notes.py

**Comment by eliben** [comment]
Thanks; I left some comments. Also, are there tests? Some of the binaries in the readelf-testing harness should have notes sections, maybe?

File: elftools/elf/segments.py (hunk 2 above)
Comment (gbenson) [inline elftools/elf/segments.py]: Ok, I'll do that.

**Comment by gbenson** [comment]
Thanks for your suggestions. I've refactored it into `elftools/elf/notes.py` as you suggested, and I've written an example which the testsuite executes.

File: elftools/elf/sections.py [hunk 3]
```diff
@@ -174,3 +176,16 @@ def iter_symbols(self):
...
+
+class NoteSection(Section):
+    """ ELF NOTE section. Knows how to parse notes.
+    """
+    def __init__(self, header, name, stream, elffile):
+        super(NoteSection, self).__init__(header, name, stream)
+        self.elffile = elffile
+
+    def iter_notes(self):
+        """ Iterates the list of notes in the segment.
```
Comment (eliben) [inline elftools/elf/sections.py]: s/segment/section/ here

File: elftools/elf/notes.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+# Eli Bendersky (eliben@gmail.com)
+# This code is in the public domain
+#-------------------------------------------------------------------------------
+from ..common.py3compat import bytes2str
+from ..common.utils import struct_parse, roundup
+from ..construct import CString
+
+
+def iter_notes(elffile, offset, size):
+    """ Iterates the list of notes in a section or segment.
```
Comment (eliben) [inline elftools/elf/notes.py]: Can this comment briefly state what is actually yielded on each iteration?

**Comment by eliben** [comment]
Thanks, looks much better. I added a couple more minor comments and it's good to go (please squash again before I can pull)

**Comment by gbenson** [comment]
Is this ok?

**Pull request merged**